core/src/main/java/com/netflix/conductor/core/execution/tasks/ExclusiveJoin.java

```diff
@@ -0,0 +1,109 @@
+	private static Logger logger = LoggerFactory.getLogger(ExclusiveJoin.class);
+	private static final String NAME = "EXCLUSIVE_JOIN";
+	private static final String DEFAULT_EXCLUSIVE_JOIN_TASKS = "defaultExclusiveJoinTask";
+
+	public ExclusiveJoin() {
+		super(NAME);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean execute(Workflow workflow, Task task, WorkflowExecutor provider) {
+
+		boolean foundExlusiveJoinOnTask = false;
+		boolean hasFailures = false;
+		boolean useWorkflowInput = false;
+		StringBuilder failureReason = new StringBuilder();
+		Task.Status taskStatus = null;
+		List<String> joinOn = (List<String>) task.getInputData().get("joinOn");
+		Task exclusiveTask = null;
+		for (String joinOnRef : joinOn) {
+			logger.info(" Exclusive Join On Task {} ", joinOnRef);
+			exclusiveTask = workflow.getTaskByRefName(joinOnRef);
+			if (exclusiveTask == null || exclusiveTask.getStatus() == Task.Status.SKIPPED) {
+				logger.debug("The task {} is either not scheduled or skipped.", joinOnRef);
+				continue;
+			}
+			taskStatus = exclusiveTask.getStatus();
+			foundExlusiveJoinOnTask = taskStatus.isTerminal();
+			hasFailures = !taskStatus.isSuccessful();
+			if (hasFailures) {
+				failureReason.append(exclusiveTask.getReasonForIncompletion()).append(" ");
+			}
+
+			break;
+		}
+
+		if (!foundExlusiveJoinOnTask) {
+			List<String> defaultExclusiveJoinTasks = (List<String>) task.getInputData()
+					.get(DEFAULT_EXCLUSIVE_JOIN_TASKS);
+			logger.info(
+					"Could not perform exclusive on Join Task(s). Performing now on default exclusive join task(s) {}",
+					defaultExclusiveJoinTasks);
+			if (defaultExclusiveJoinTasks != null && !defaultExclusiveJoinTasks.isEmpty()) {
+				for (String defaultExclusiveJoinTask : defaultExclusiveJoinTasks) {
+					// Pick the first task that we should join on and break.
+					exclusiveTask = workflow.getTaskByRefName(defaultExclusiveJoinTask);
+					if (exclusiveTask == null || exclusiveTask.getStatus() == Task.Status.SKIPPED) {
+						logger.debug("The task {} is either not scheduled or skipped.", defaultExclusiveJoinTask);
+						continue;
+					}
+
+					taskStatus = exclusiveTask.getStatus();
+					foundExlusiveJoinOnTask = taskStatus.isTerminal();
+					hasFailures = !taskStatus.isSuccessful();
+					if (hasFailures) {
+						failureReason.append(exclusiveTask.getReasonForIncompletion()).append(" ");
+					}
+					break;
+
+				}
+			} else {
+				useWorkflowInput = true;
+			}
+		}
+
+		logger.debug("Status of flags: foundExlusiveJoinOnTask: {}, hasFailures {}, useWorkflowInput {}",
+				foundExlusiveJoinOnTask, hasFailures, useWorkflowInput);
+		if (foundExlusiveJoinOnTask || hasFailures || useWorkflowInput) {
+			if (hasFailures) {
+				task.setReasonForIncompletion(failureReason.toString());
+				task.setStatus(Task.Status.FAILED);
+			} else {
+				task.setOutputData((useWorkflowInput) ? workflow.getInput() : exclusiveTask.getOutputData());
```

Comment:
Can you explain why workflow input is wired as the output of this task if an exclusive task is not defined?

Reply:
By contract exclusive join has to provide the previous tasks output. Proper value is expected as exclusive task or the default task. If neither is matched, the workflows input is passed as is as output to drive the next task in the sequence.  Do you think it is a better idea to have a configuration flag that determines if this task has to consider workflow input as the output?

Reply:
In this case, the output of the task should be `null` or `empty` to explicitly indicate that no cases matched and the default fallback was used. The workflow input is accessible from within any task as-is and does not need this additional indirection.

Reply:
Incorporated review comments.